src/common/concurrent/dlock.h

```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright (c) 2020 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: April 01 2021
+ * Author: wanghai01
+ */
+
+#ifndef SRC_COMMON_CONCURRENT_DLOCK_H_
+#define SRC_COMMON_CONCURRENT_DLOCK_H_
+
+#include <string>
+#include <memory>
+#include "src/kvstorageclient/etcd_client.h"
+#include "src/common/uncopyable.h"
+
+
+namespace curve {
+namespace common {
+
+using curve::kvstorage::KVStorageClient;
+using curve::common::Uncopyable;
+
+struct DLockOpts {
+    std::string pfx;
+    int retryTimes;
+    // the interface timeout，unit is millisecond
+    int timeout;
+    // the session and lease timeout, unit is second
+    int ttl;
+};
+
+
+class DLock : public Uncopyable {
+ public:
+    explicit DLock(const DLockOpts &opts) : opts_(opts), locker_(0) {}
+    virtual ~DLock();
+
+    /**
+     * @brief Init the etcd Mutex
+     * 
+     * @return error code EtcdErrCode
+     */
+    virtual int64_t Init();
+
+    /**
+     * @brief lock the object
+     * 
+     * @return error code EtcdErrCode
+     */
+    virtual int Lock();
+
+    /**
+     * @brief try to lock the object
+     *
+     * @return error code EtcdErrCode
+     */
+    // int TryLock();
+
+    /**
+     * @brief unlock the object
+     *
+     * @return error code EtcdErrCode
+     */
+    virtual int Unlock();
+
+ private:
+    bool NeedRetry(int errCode);
+
+ private:
+    int64_t locker_;
+    const DLockOpts &opts_;
+};
+
+class DLockGuard : public Uncopyable {
+ public:
+    explicit DLockGuard(std::shared_ptr<DLock> lock): lock_(lock) {
+        lock_->Lock();
+    }
+
+    ~DLockGuard() {
```

Comment:
> Do I have to manually call Release()?
 
yes.
after set dlock in task's closure Release, and Unlock in call backfun
DLockGuard dlockGuard(dlock_);
closure->SetDLock(dlock_);
dlockGuard.Release();

Reply:
Does Releaese() used to avoid repeated destruction?

Reply:
> Does Releaese() used to avoid repeated destruction?

the Release() is used to not destructed and give the unlock to task callback
~DLockGuard() {
        if (!release_) {
            lock_->Unlock();
        }
    }